xayn-ai/src/coi/system.rs

```diff
@@ -214,6 +223,160 @@ impl CoiSystem {
+                    .then(|| {
+                        smbert(words)
+                            .ok()
+                            .and_then(|point| KeyPhrase::new(words, point).ok())
+                    })
+                    .flatten()
+            })
+            .collect::<BTreeSet<_>>();
+
+        let len = coi.key_phrases().len() + candidates.len();
+        let max_key_phrases = self.config.max_key_phrases.min(len);
+        if max_key_phrases == 0 {
+            coi.swap_key_phrases(BTreeSet::default());
+            return;
+        }
+
+        let similarity = pairwise_cosine_similarity(
+            coi.key_phrases()
+                .iter()
+                .chain(candidates.iter())
+                .map(|key_phrase| key_phrase.point().view())
+                .chain(once(coi.point().view())),
+        )
+        .slice_move(s![..len, ..]);
+        debug_assert!(similarity.iter().copied().all(f32::is_finite));
+
+        let min = reduce_without_diag(similarity.view(), Axis(0), f32::min, identity);
+        let max = reduce_without_diag(similarity.view(), Axis(0), f32::max, identity);
+        let normalized = (&similarity - &min) / (max - min);
+        let mean = reduce_without_diag(
+            normalized.view(),
+            Axis(0),
+            |reduced, element| reduced + element,
+            |reduced| reduced / (len - 1) as f32,
+        );
+        let std_dev = reduce_without_diag(
+            &normalized - &mean,
+            Axis(0),
+            |reduced, element| reduced + element.powi(2),
+            |reduced| (reduced / (len - 1) as f32).sqrt(),
+        );
+        let normalized = (normalized - mean) / std_dev + 0.5;
+        let normalized = normalized
+            .mapv_into(|normalized| normalized.is_finite().then(|| normalized).unwrap_or(0.5));
+        debug_assert!(normalized.iter().copied().all(f32::is_finite));
+
+        let candidate = argmax(normalized.slice(s![.., len]));
+        let mut selected = vec![false; len];
+        selected[candidate] = true;
+        for _ in 0..max_key_phrases - 1 {
+            let candidate = argmax(selected.iter().zip(normalized.rows()).map(
+                |(&is_selected, normalized)| {
+                    if is_selected {
+                        f32::MIN
+                    } else {
+                        let max = selected
+                            .iter()
+                            .zip(normalized)
+                            .filter_map(|(is_selected, normalized)| {
+                                is_selected.then(|| *normalized)
+                            })
+                            .reduce(f32::max)
+                            .unwrap(/* at least one key phrase is selected */);
+                        self.config.gamma * normalized[len] - (1. - self.config.gamma) * max
+                    }
+                },
+            ));
+            selected[candidate] = true;
+        }
+
+        let relevance = selected
```

Comment:
if we get the selected index:
```rust
let selected = selected.enumerate().filter(|(_, selected)| selected).collect();
let relevance = similarity.select(Axis(0), selected);
```

Reply:
i didn't know about the `select()` method, that can be quite useful 👍 in this case i would still keep the current code, because your suggestion adds an additional vector allocation, whereas the current code allocates only once, because line 357 returns an iterator and not a vector.